### DISCUSSION TITLE
Question concerning the playback feature

### DISCUSSION THREAD

**Discussion by dstrn825** [rany2/edge-tts#87] [Q&A] [answered]
Hello,
I've been trying to use the playback feature in my Python script as I'm trying to make a TTS program.
How can I use the playback feature of the library? I don't want the output to be saved in a file but rather directly play the audio. I'd also like to be able to edit the pitch but couldn't find any example that does this amongst the provided ones.
Thanks in advance.

**Comment by rany2** [comment, marked as answer]
All the audio playback libraries I know eventually would put the data on disk, one way or the other even when you didn't explicitly ask for it. Really just putting it in tmpfs should be OK, I'm not sure why everyone has an issue with that. 

Also the data received from server is either WebM or MP3 and both would need to be converted before being sent to the audio server, and the only ways available to convert these to formats supported by the audio server would require you to put the data in a file.

**Comment by xiaoqiangclub** [comment]
@dstrn825 Hello, I also have the same need as you, want to achieve direct playback of audio streams without saving mp3 files, I don't know if you have solved this problem, if you have solved it, can you share your method, thank you

**Reply by dstrn825** [reply]
Yo, yeah I fixed both the pitch and playback issue. Basically I save the audio data in a temporary wav file, then load the audio data in a variable, delete the temp file, and change the pitch with a loop I made. After that I save a new temp file and play it back before deleting it again (you can playback on speaker or whatever device you want, in my case I play it in discord calls).
To save the file, I just used this function and asyncio to save the file: 
```
async def save(communicate):
    await communicate.save(FILE)
asyncio.run(save(communicate))
```
To play the file, I used the _sounddevice_ and _soundfile_ libraries. You can change the playback device etc.
You can then easily delete the temp file by using `os.remove(FILE)`
Seeing your question I assume you weren't referring to the pitching part so if all you care about is the playback aspect, you won't even notice the temp file being created as it will be instantly deleted after.
Lemme know if you need more help or add on disc if you wish
